src/crypto.c

```diff
@@ -185,7 +182,7 @@ zxerr_t crypto_sign(const hd_path_t path,
                                sizeof(messageDigest),
                                &messageDigestSize));
 
-    if (messageDigestSize != CX_SHA256_SIZE) {
+    if (messageDigestSize != CURVE_ORDER_SIZE) {
```

Comment:
In theory, it's enough to only check `messageDigestSize < CURVE_ORDER_SIZE` to error. `messageDigestSize >= CURVE_ORDER_SIZE` is the allowed ECDSA requirement, but that also depends on how `bip32_derive_ecdsa_sign_hash_256` is implemented internally. In Flow, the supported hashing output was forced to match the curve order size, hence the check for equality here.

On a slightly different note, if `bip32_derive_ecdsa_sign_hash_256` checks for the curve/digest compatibility internally, then this check can be reduced to checking that hashing has worked correctly, as in `messageDigestSize` is equal to `CX_SHA256_SIZE` or `CX_SHA3_256_SIZE`.

Reply:
After checking the [doc](https://github.com/LedgerHQ/ledger-secure-sdk/blob/459c14bcdff748d1eac0204921cb06b90a73fa06/lib_standard_app/crypto_helpers.h#L203) of `bip32_derive_ecdsa_sign_hash_256`, I see that there is no check of compatibility of the group order/hash length!

The caller of `bip32_derive_ecdsa_sign_hash_256` needs to make the check in this case. My preference in this case is to: 
 1. explicitly check that the hashing worked as expected, as in the digest size is equal to the expected hash output constant
 2. explicitly check that the digest size is at least the curve group order (not checked by the Ledger SDK).
 
If we skip (2), the code may integrate for instance sha224 and use it to sign without an error. The security of the curve (128 bits) is silently reduced in this case to the security of the hash (112 bits), which is not good. Let me know if you have questions to clarify

Reply:
check added https://github.com/vacuumlabs/app-flow/pull/4/commits/1e5abf83528fe276dc6ce035389577c768d48eb6